### PR TITLE
Bigtable: 01. Surface: Base structure

### PR DESCRIPTION
**Pull request by igorbernstein2**
Layout the base structure to overlay the Bigtable GAPIC client. The general idea is for EnhancedBigtableStub to wrap the GAPIC GrpcBigtableStub
and for it to construct Callable chains on top of the generated chains.

This has been reviewed Kevin. However some things have changed since the original review:
- With the introduction of StubSettings all of the configuration has been moved to a new `EnhancedBigtableStubSettings`
- `BigtableDataSettings` has been updated to just delegate to `EnhancedBigtableStubSettings`
- `EnhancedBigtableStub` used to be called `BigtableDataStub`, but was renamed to avoid ambiguity

### CODE REVIEW

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * terminates in a Callable from {@link GrpcBigtableStub}. This class is meant to be a semantically
+ * complete facade for the Bigtable data API. However it is not meant to be consumed directly,
+ * please use {@link com.google.cloud.bigtable.data.v2.BigtableDataClient}.
+ *
+ * <p>This class is considered an internal implementation detail and not meant to be used by
+ * applications.
+ */
+@InternalApi
+public class EnhancedBigtableStub implements AutoCloseable {
+  private final EnhancedBigtableStubSettings settings;
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Please ignore, this PR is setting up structure that will be used in upcoming PRs

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  private final GrpcBigtableStub stub;
+  private final ClientContext clientContext;
+  private final RequestContext requestContext;
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Please ignore, this PR is setting up structure that will be used in upcoming PRs

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * please use {@link com.google.cloud.bigtable.data.v2.BigtableDataClient}.
+ *
+ * <p>This class is considered an internal implementation detail and not meant to be used by
+ * applications.
+ */
+@InternalApi
+public class EnhancedBigtableStub implements AutoCloseable {
+  private final EnhancedBigtableStubSettings settings;
+  private final GrpcBigtableStub stub;
+  private final ClientContext clientContext;
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Please ignore, this PR is setting up structure that will be used in upcoming PRs

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Please ignore, this PR is setting up structure that will be used in upcoming PRs

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Please ignore, this PR is setting up structure that will be used in upcoming PRs

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Please ignore, this PR is setting up structure that will be used in upcoming PRs

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    // </editor-fold>
+
+    // <editor-fold desc="Public API">
+    /**
+     * Sets the target instance. This setting is required. All RPCs will be made in the context of
+     * this setting.
+     */
+    public Builder setInstanceName(InstanceName instanceName) {
+      Preconditions.checkNotNull(instanceName);
+      this.instanceName = instanceName;
```
Comment (pongad) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: You can `this.instanceName = Preconditions.checkNotNull(instanceName)`. Up to you though.

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    EnhancedBigtableStubSettings.Builder builder = EnhancedBigtableStubSettings.newBuilder();
+
+    Throwable error = null;
+    try {
+      builder.build();
+    } catch (Throwable t) {
+      error = t;
+    }
+
+    Truth.assertThat(error).isInstanceOf(IllegalStateException.class);
```
Comment (pongad) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java]: I think the recommendation is to `import static com.google.common.truth.Truth.assertThat` and just use `assertThat` in the tests.

**Comment by igorbernstein2** [comment]
All feedback has been incorporated. Also some minor changes applied:
- code reformatted via googleformat maven plugin
- Added Nonnull annotations to set{InstanceName,AppProfile}
- Suppressed unchecked warning
- Removed stale checked exception

Please merge when ready

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Please ignore, this PR is setting up structure that will be used in upcoming PRs

**Comment by pongad** [comment]
LGTM. @garrettjonesgoogle PTAL

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+ * build() is called, the tree of builders is called to create the complete settings object.
+ *
+ * <pre>{@code
+ * BigtableDataSettings bigtableDataSettings = BigtableDataSettings.newBuilder()
+ *   .setInstanceName(InstanceName.of("my-project", "my-instance-id"))
+ *   .setAppProfileId("default")
+ *   .build();
+ * }</pre>
+ */
+public class EnhancedBigtableStubSettings extends StubSettings<EnhancedBigtableStubSettings> {
```
Comment (garrettjonesgoogle) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Shouldn't this be `@InternalApi` like `EnhancedBigtableStub`?

**Comment by igorbernstein2** [comment]
Good catch! Updated, please merge

**Pull request merged**